File: client/src/lib/first-run.ts
import type { User } from "@pqp/shared";

/**
 * The first-run checklist's decisions, with no React attached.
 *
 * Same split as `onboarding.ts`, for the same reason: the component is a list of
 * three rows and two buttons, and everything that can be got *wrong* — who sees
 * it, who never sees it again, and when "done" is done — is here where it can be
 * tested without a DOM.
 *
 * WHAT THIS FIXES. The wizard ends and the app opens on the hub, which for an
 * account with no servers is the Friends view with "No friends here yet" in it.
 * That is one of the three things a new account needs, and it is the only one
 * the hub ever mentions: `Create server` and `Join invite` live in the empty
 * state for a *selected server*, which an account with no servers cannot reach,
 * and the avatar is three clicks into a settings modal nothing points at. So the
 * screen a new person lands on after signing up offers one third of what they
 * came to do, and the other two thirds are unlabelled icons in a 56px rail.
 */

/** The three things, in the order they unlock each other. */
export type FirstRunTaskId = "server" | "friend" | "avatar";

export interface FirstRunTask {
  id: FirstRunTaskId;
  done: boolean;
}

export interface FirstRunState {
  /** All three, in display order, each with its own done-ness. */
  tasks: FirstRunTask[];
  /** Nothing left to do. */
  complete: boolean;
}

export interface FirstRunInputs {
  user: User | null;
  serverCount: number;
  friendCount: number;
}

/**
 * What is still outstanding, read from live state rather than from stored flags.
 *
 * Derived on purpose. A stored "created a server" bit would go stale the moment
 * somebody leaves the server again, and worse, it would need a write on every
 * one of the three paths — three more chances to record something that did not
 * happen. Counting what the app already has loaded cannot drift from the truth.
 *
 * `avatarUrl` is the whole avatar test. An empty string counts as absent: the
 * profile step stores `null` when the field is cleared, but a PATCH that trims
 * to nothing has historically round-tripped as `""`, and a checklist that calls
 * that done would tick a box for a face nobody can see.
 */
export function firstRunState({
  user,
  serverCount,
  friendCount,
}: FirstRunInputs): FirstRunState {
  const tasks: FirstRunTask[] = [
    { id: "server", done: serverCount > 0 },
    { id: "friend", done: friendCount > 0 },
    { id: "avatar", done: Boolean(user?.avatarUrl) },
  ];
  return { tasks, complete: tasks.every((task) => task.done) };
}

/**
 * Should the hub draw the checklist?
 *
 * Three "no"s, and they are the whole function:
 *
 *  - `preferences` entirely absent — an API that cannot record a dismissal.
 *    Offering a card whose "no thanks" does not stick means offering it forever,
 *    which is the nagging this was built to avoid. Same reasoning
 *    `shouldRunOnboarding` applies to the same missing field.
 *  - `firstRunDismissedAt` present — answered. Never again, on any device.
 *  - everything done — there is nothing to offer.
 *
 * NOTHING HERE LOOKS AT `onboardedAt`, and that is not an omission. It reads as
 * the obvious fourth guard — "do not stack a checklist under the wizard" — but
 * the wizard is a full-screen replacement, so the hub cannot render while it is
 * up and the guard protects against nothing. What it *would* do is suppress the
 * card in the one session it exists for: the wizard writes `onboardedAt` to the
 * server without patching the app's copy of the user (see `finish()` in
 * `onboarding-flow.tsx` for why it must not), so for the rest of that session the
 * local user still says onboarding never ran. Gating on it meant the person who
 * had just signed up — the entire audience — saw nothing.
 */
export function shouldShowFirstRun(
  inputs: FirstRunInputs,
  state: FirstRunState = firstRunState(inputs),
): boolean {
  const preferences = inputs.user?.preferences;
  if (!preferences) {
    return false;
  }
  if (preferences.firstRunDismissedAt) {
    return false;
  }
  return !state.complete;
}

/**
 * Should the card's dismissal be recorded without anybody clicking anything?
 *
 * Yes exactly when there is nothing left to do and nothing has been recorded.
 * This is what closes the derived-state loop described in the schema comment:
 * the card stops rendering the moment the third item is done, but "stops
 * rendering" is not "gone" while the reason it stopped can be undone. One write,
 * once per account, and the question can never be asked again.
 *
 * Deliberately does not require `onboardedAt`. An account still inside the
 * wizard that already has all three (joined by invite, has a friend, uploaded an
 * avatar in step 2) is as finished as any other, and making it wait would mean
 * asking it on the next cold start instead.
 */
export function shouldStampFirstRunComplete(
  inputs: FirstRunInputs,
  state: FirstRunState = firstRunState(inputs),
): boolean {
  const preferences = inputs.user?.preferences;
  if (!preferences || preferences.firstRunDismissedAt) {
    return false;
  }
  return state.complete;
}

/** The preference patch that closes the checklist for good, on every device. */
export function firstRunDismissedPatch(now: Date = new Date()): {
  firstRunDismissedAt: string;
} {
  return { firstRunDismissedAt: now.toISOString() };
}

File: ios/pqp/Sources/Onboarding/FirstRun.swift
import Foundation

/// The first-run checklist's decisions, with no SwiftUI attached.
///
/// The mirror of `client/src/lib/first-run.ts`, and deliberately so: the flag it
/// reads is a *server* preference, so web and iOS have to agree on when the card
/// shows and when it is answered, or dismissing it on a phone would leave it
/// sitting there on a laptop. Keeping the rules in a plain value type also means
/// they can be tested without a simulator, the way `LastVisited` is.
///
/// WHAT IT FIXES ON iOS. The hub is better off than the web's was — it has a real
/// "Create a server" card and a real "Start a conversation" card. What it has
/// never had is any pointer at *friends* (the only affordance is an unlabelled
/// `person.2.fill` circle in the dock, with no badge at zero) or at an avatar
/// (three taps deep behind that dock, and nothing in first run mentions it). And
/// "Start a conversation" on a fresh account opens a sheet whose suggestion list
/// is friends-only, so it opens on nothing at all.
enum FirstRunTask: String, CaseIterable, Identifiable, Sendable {
    case server
    case friend
    case avatar

    var id: String { rawValue }
}

/// What is still outstanding, and whether that is nothing.
struct FirstRunState: Equatable, Sendable {
    /// Every task, in display order, paired with its done-ness.
    let tasks: [(task: FirstRunTask, done: Bool)]

    var complete: Bool { tasks.allSatisfy(\.done) }

    func isDone(_ task: FirstRunTask) -> Bool {
        tasks.first { $0.task == task }?.done ?? false
    }

    static func == (lhs: FirstRunState, rhs: FirstRunState) -> Bool {
        lhs.tasks.count == rhs.tasks.count
            && zip(lhs.tasks, rhs.tasks).allSatisfy {
                $0.task == $1.task && $0.done == $1.done
            }
    }
}

enum FirstRun {
    /// Everything the two questions below are asked about.
    struct Inputs: Sendable {
        /// Nil where the account has not loaded yet.
        var avatarURL: String?
        var serverCount: Int
        var friendCount: Int
        /// Nil where the API predates the preference store entirely.
        var preferences: UserPreferences?

        init(
            avatarURL: String? = nil,
            serverCount: Int = 0,
            friendCount: Int = 0,
            preferences: UserPreferences? = UserPreferences()
        ) {
            self.avatarURL = avatarURL
            self.serverCount = serverCount
            self.friendCount = friendCount
            self.preferences = preferences
        }
    }

    /// What is outstanding, read from live state rather than from stored flags.
    ///
    /// Derived on purpose, exactly as on the web. A stored "created a server" bit
    /// would go stale the moment somebody leaves the server again, and it would
    /// need a write on each of the three paths — three more chances to record
    /// something that did not happen.
    ///
    /// An empty avatar string counts as absent: a cleared field has round-tripped
    /// as `""` rather than null, and ticking that box would credit somebody for a
    /// face nobody can see.
    static func state(_ inputs: Inputs) -> FirstRunState {
        let hasAvatar = !(inputs.avatarURL ?? "").isEmpty
        return FirstRunState(tasks: [
            (.server, inputs.serverCount > 0),
            (.friend, inputs.friendCount > 0),
            (.avatar, hasAvatar),
        ])
    }

    /// Should the hub draw the checklist?
    ///
    /// Three "no"s: no preference store to record a dismissal in (a card whose
    /// "no thanks" does not stick is a card that nags forever), already dismissed,
    /// or nothing left to offer.
    ///
    /// Nothing here consults an onboarding flag. iOS onboarding is three marketing
    /// beats and a sign-in, recorded in a device-local `UserDefaults` bool that
    /// says nothing about whether this account has a server, a friend or a face.
    static func shouldShow(_ inputs: Inputs) -> Bool {
        guard let preferences = inputs.preferences else { return false }
        guard !isDismissed(preferences) else { return false }
        return !state(inputs).complete
    }

    /// Has the checklist been answered?
    ///
    /// An EMPTY STRING COUNTS AS NOT ANSWERED, and that is not a technicality.
    /// Preferences are one JSONB blob merged with `||` server-side, so a key can
    /// be overwritten but never removed — which means once this field is written
    /// there is no way back to absent, and no way to put an account back in the
    /// state that shows the card. Reading `""` as "never dismissed" is the only
    /// re-arm there is, and it is what makes the flag testable at all.
    ///
    /// The web reaches the same answer through JavaScript truthiness
    /// (`if (preferences.firstRunDismissedAt)`), where `""` is already falsy. This
    /// has to say it out loud, because in Swift `""` is not nil.
    static func isDismissed(_ preferences: UserPreferences) -> Bool {
        !(preferences.firstRunDismissedAt ?? "").isEmpty
    }

    /// Should the dismissal be recorded without anybody tapping anything?
    ///
    /// Yes exactly when there is nothing left to do and nothing recorded. This is
    /// what makes "never returns" true rather than "usually does not return":
    /// visibility is derived from live state, and live state comes back. Leaving
    /// your last server a year from now must not reopen a "get into a server"
    /// nudge at somebody who has been here a year.
    static func shouldStampComplete(_ inputs: Inputs) -> Bool {
        guard let preferences = inputs.preferences else { return false }
        guard !isDismissed(preferences) else { return false }
        return state(inputs).complete
    }

    /// The instant to store, formatted the way the shared schema expects.
    ///
    /// `onboardedAt` and this are both bare `z.string()` on the server, so nothing
    /// validates the format — which is exactly why it is pinned here rather than
    /// left to whatever `Date.description` happens to produce. ISO-8601 with a
    /// `Z`, matching what `new Date().toISOString()` writes from the web, so a row
    /// dismissed on a phone reads the same as one dismissed in a browser.
    static func dismissedStamp(_ now: Date = Date()) -> String {
        let formatter = ISO8601DateFormatter()
        formatter.timeZone = TimeZone(secondsFromGMT: 0)
        formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        return formatter.string(from: now)
    }
}
